In [1]:
import pandas as pd

# Loading the data

In [5]:
users = pd.read_csv("users.csv")

In [44]:
users.shape

(495, 7)

In [11]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   _id.$oid           495 non-null    object 
 1   active             495 non-null    bool   
 2   createdDate.$date  495 non-null    int64  
 3   lastLogin.$date    433 non-null    float64
 4   role               495 non-null    object 
 5   signUpSource       447 non-null    object 
 6   state              439 non-null    object 
dtypes: bool(1), float64(1), int64(1), object(4)
memory usage: 23.8+ KB


Conclusion: lastLogin, signUpSource,state have null values

In [13]:
users.describe()

createdDate.$date  lastLogin.$date
count       4.950000e+02     4.330000e+02
mean        1.596678e+12     1.611388e+12
std         4.402970e+10     4.430109e+09
min         1.418999e+12     1.525714e+12
25%         1.609789e+12     1.610130e+12
50%         1.610569e+12     1.611237e+12
75%         1.611596e+12     1.612366e+12
max         1.613139e+12     1.614963e+12

The createdDate and lastLogin date are of datatype integer and float but the datatype of both has to be Date
and value of createdDate,lastLogin are not in the format of date

In [33]:
users.isna().sum().sum()

166

There are total of 166 null values in the "users" dataset which is around 34% of data missing

# Brands

In [7]:
brands = pd.read_csv("brands.csv")

In [45]:
brands.shape

(1167, 9)

In [34]:
brands.isna().sum().sum()

35

There are 35 null values in Brands data set

In [38]:
brands

_id.$oid       barcode            category  \
0     601ac115be37ce2ead437551  511111019862              Baking   
1     601c5460be37ce2ead43755f  511111519928           Beverages   
2     601ac142be37ce2ead43755d  511111819905              Baking   
3     601ac142be37ce2ead43755a  511111519874              Baking   
4     601ac142be37ce2ead43755e  511111319917      Candy & Sweets   
...                        ...           ...                 ...   
1162  5f77274dbe37ce6b592e90c0  511111116752              Baking   
1163  5dc1fca91dda2c0ad7da64ae  511111706328  Breakfast & Cereal   
1164  5f494c6e04db711dd8fe87e7  511111416173      Candy & Sweets   
1165  5a021611e4b00efe02b02a57  511111400608             Grocery   
1166  6026d757be37ce6369301468  511111019930              Baking   

          categoryCode              cpg.$id.$oid cpg.$ref  \
0               BAKING  601ac114be37ce2ead437550     Cogs   
1            BEVERAGES  5332f5fbe4b03c9a25efd0ba     Cogs   
2               BAKING  601ac142be37ce2ead437559     Cogs   
3               BAKING  601ac142be37ce2ead437559     Cogs   
4     CANDY_AND_SWEETS  5332fa12e4b03c9a25efd1e7     Cogs   
...                ...                       ...      ...   
1162            BAKING  5f77274dbe37ce6b592e90bf     Cogs   
1163                ""  53e10d6368abd3c7065097cc     Cogs   
1164  CANDY_AND_SWEETS  5332fa12e4b03c9a25efd1e7     Cogs   
1165                ""  5332f5f6e4b03c9a25efd0b4     Cogs   
1166            BAKING  6026d757be37ce6369301467     Cogs   

                           name topBrand                      brandCode  
0     test brand @1612366101024    false                             ""  
1                     Starbucks    false                      STARBUCKS  
2     test brand @1612366146176    false  TEST BRANDCODE @1612366146176  
3     test brand @1612366146051    false  TEST BRANDCODE @1612366146051  
4     test brand @1612366146827    false  TEST BRANDCODE @1612366146827  
...                         ...      ...                            ...  
1162  test brand @1601644365844       ""                             ""  
1163        Dippin Dots® Cereal       ""             DIPPIN DOTS CEREAL  
1164  test brand @1598639215217       ""  TEST BRANDCODE @1598639215217  
1165          LIPTON TEA Leaves    false              LIPTON TEA Leaves  
1166  test brand @1613158231643    false  TEST BRANDCODE @1613158231644  

[1167 rows x 9 columns]

In [37]:
brands["topBrand"].value_counts()

""       612
false    524
true      31
Name: topBrand, dtype: int64

The topBrand column has "612" values which are stored as ""(empty values) and datatype should be Boolean
For this column around 52% of the data is missing which is a major concern

In [40]:
brands["brandCode"].value_counts()

""                               234
HUGGIES                            2
GOODNITES                          2
PURE BLISS                         1
TEST BRANDCODE @1602694020787      1
                                ... 
WHEATIES                           1
FREIHOFERS                         1
OVEN FRY                           1
MAISER'S                           1
TEST BRANDCODE @1610562208680      1
Name: brandCode, Length: 897, dtype: int64

The brandCode column has "234" values which are stored as ""(empty values) and datatype should be Boolean

# Receipts

receipts = pd.read_csv("receipts.csv")

In [10]:
receipts

_id.$oid bonusPointsEarned  \
0     5ff1e1eb0a720f0523000575               500   
1     5ff1e1bb0a720f052300056b               150   
2     5ff1e1f10a720f052300057a                 5   
3     5ff1e1ee0a7214ada100056f                 5   
4     5ff1e1d20a7214ada1000561                 5   
...                        ...               ...   
1114  603cc0630a720fde100003e6                25   
1115  603d0b710a720fde1000042a                ""   
1116  603cf5290a720fde10000413                ""   
1117  603ce7100a7217c72c000405                25   
1118  603c4fea0a7217c72c000389                ""   

                                bonusPointsEarnedReason  createDate.$date  \
0     Receipt number 2 completed, bonus point schedu...     1609687531000   
1     Receipt number 5 completed, bonus point schedu...     1609687483000   
2                            All-receipts receipt bonus     1609687537000   
3                            All-receipts receipt bonus     1609687534000   
4                            All-receipts receipt bonus     1609687506000   
...                                                 ...               ...   
1114                        COMPLETE_NONPARTNER_RECEIPT     1614594147000   
1115                                                 ""     1614613361873   
1116                                                 ""     1614607657664   
1117                        COMPLETE_NONPARTNER_RECEIPT     1614604048000   
1118                                                 ""     1614565354962   

      dateScanned.$date finishedDate.$date  modifyDate.$date  \
0         1609687531000      1609687531000     1609687536000   
1         1609687483000      1609687483000     1609687488000   
2         1609687537000                 ""     1609687542000   
3         1609687534000      1609687534000     1609687539000   
4         1609687506000      1609687511000     1609687511000   
...                 ...                ...               ...   
1114      1614594147000                 ""     1614594148000   
1115      1614613361873                 ""     1614613361873   
1116      1614607657664                 ""     1614607657664   
1117      1614604048000                 ""     1614604049000   
1118      1614565354962                 ""     1614565354962   

     pointsAwardedDate.$date pointsEarned purchaseDate.$date  \
0              1609687531000        500.0      1609632000000   
1              1609687483000        150.0      1609601083000   
2                         ""            5      1609632000000   
3              1609687534000          5.0      1609632000000   
4              1609687506000          5.0      1609601106000   
...                      ...          ...                ...   
1114                      ""         25.0      1597622400000   
1115                      ""           ""                 ""   
1116                      ""           ""                 ""   
1117                      ""         25.0      1597622400000   
1118                      ""           ""                 ""   

     purchasedItemCount                             rewardsReceiptItemList  \
0                     5  [{"barcode":"4011","description":"ITEM NOT FOU...   
1                     2  [{"barcode":"4011","description":"ITEM NOT FOU...   
2                     1  [{"needsFetchReview":false,"partnerItemId":"1"...   
3                     4  [{"barcode":"4011","description":"ITEM NOT FOU...   
4                     2  [{"barcode":"4011","description":"ITEM NOT FOU...   
...                 ...                                                ...   
1114                  2  [{"barcode":"B076FJ92M4","description":"muelle...   
1115                 ""                                                 ""   
1116                 ""                                                 ""   
1117                  2  [{"barcode":"B076FJ92M4","description":"muelle...   
1118                 ""                                                 ""  

In [41]:
receipts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   _id.$oid                 1119 non-null   object
 1   bonusPointsEarned        1119 non-null   object
 2   bonusPointsEarnedReason  1119 non-null   object
 3   createDate.$date         1119 non-null   int64 
 4   dateScanned.$date        1119 non-null   int64 
 5   finishedDate.$date       1119 non-null   object
 6   modifyDate.$date         1119 non-null   int64 
 7   pointsAwardedDate.$date  1119 non-null   object
 8   pointsEarned             1119 non-null   object
 9   purchaseDate.$date       1119 non-null   object
 10  purchasedItemCount       1119 non-null   object
 11  rewardsReceiptItemList   1119 non-null   object
 12  rewardsReceiptStatus     1119 non-null   object
 13  totalSpent               1119 non-null   object
 14  userId                   1119 non-null  

In [23]:
receipts['bonusPointsEarned'].value_counts()

""     575
5      183
750    119
25      71
250     31
45      31
500     30
150     27
300     26
100     18
27       6
21       1
40       1
Name: bonusPointsEarned, dtype: int64

The bonusPointsEarned column has "575" values which are stored as ""(empty values) 

In [24]:
receipts['bonusPointsEarnedReason'].value_counts()

""                                                                                     575
All-receipts receipt bonus                                                             183
Receipt number 1 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)    119
COMPLETE_NONPARTNER_RECEIPT                                                             71
COMPLETE_PARTNER_RECEIPT                                                                39
Receipt number 3 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)     31
Receipt number 2 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)     30
Receipt number 5 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)     27
Receipt number 4 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)     26
Receipt number 6 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)     18
Name: bonusPointsEarnedReason, dtype: int64

The bonusPointsEarnedReason column has "575" values which are stored as ""(empty values) 

# Data quality Issues

I performed Exploratory Data Analysis on the three datasets and found the following data quality issues
1. There are many null values in the dataset which contributes around 33% to 52% of total data. For dataset of small size,
   it is alarming to have null values in majority and these issues has to be addressed, decisions has to be taken to deal with
   the missing values
2.The datatype are not correct(i. For dates the datatypes is int but it has to be Date ii. Boolean columns has datatypes as int)
3. Values in Date columns are incorrect, the data should be in format YYYY-MM-DD or MM-DD-YYYY 
4. Some columns have "" as values which are to be considered as null values